web-server/sgx-wallet-impl/src/ported/kv_store/mod.rs

```diff
@@ -0,0 +1,69 @@
+//! Simple key-value store abstraction
+//!
+//! XXX port note: Key type changed from str to `[u8]`.
+
+pub mod fs;
+
+pub type Key = [u8];
+
+/// A key-value store.
+///
+/// These methods borrow key and value references,
+/// to suit cloning / serialising implementations.
+///
+pub trait KvStore<V> {
+    type Error;
+
+    /// Load the saved value for `key`, if any.
+    ///
+    /// Return [`None`] if `key` has no previous value.
+    ///
+    fn load(&self, key: &Key) -> Result<Option<V>, Self::Error>;
+
+    /// Save a new value for `key`.
+    ///
+    /// This will replace any existing value.
+    ///
+    fn save(&mut self, key: &Key, value: &V) -> Result<(), Self::Error>;
+
+    /// Delete the saved value for `key`.
+    fn delete(&mut self, key: &Key) -> Result<(), Self::Error>;
+
+    /// Alter the value of `key`.
+    ///
+    /// This operation is a generalisation of [`Self::load`], [`Self::save`], and [`Self::delete`].
+    ///
+    fn alter<F>(&mut self, key: &Key, alter_fn: F) -> Result<Option<V>, Self::Error>
+    where
+        F: FnOnce(Option<V>) -> Option<V>,
+    {
+        let loaded: Option<V> = self.load(key)?;
+        let altered: Option<V> = alter_fn(loaded);
+        match &altered {
+            None => self.delete(key)?,
+            Some(value) => self.save(key, value)?,
+        };
+        Ok(altered)
+    }
+
+    /// Mutate the value of `key`.
+    ///
+    /// This is like [`Self::mutate`], but only operates on existing values.
+    fn mutate<F>(&mut self, key: &Key, mutate_fn: F) -> Result<Option<V>, Self::Error>
```

Comment:
Signature wise, it might make sense for this to return `Result<V, Self::Error>` instead, depending on if we cater for deletion or not. If I am reading this correctly, this would return None in when trying to mutate a non existing keypair. I think in that case an Error might be more predictable.

I think it helps to be clear about mutations failing in this context.

Reply:
The way the `Self::Error` type is parameterised is intended more for exceptional failure conditions (particularly IO), so we can't really use it here.

We could introduce a separate, but I don't think that gives us much semantic advantage over returning the result of the operation directly as an `Option`: this is more consistent with how the rest of the API operates.

(We could probably document it better later, but the intent of this method is not "mutate and assert that the key was present", but rather "mutate if it was present, and return whether it was mutated (along with the result, if it was)".)

Reply:
(Resolving this for now: we should probably defer bigger changes to this ported code until we've split it out to a shared library.)